keystore/filesystem/server_keystore.go

```diff
@@ -774,6 +793,75 @@ func (store *KeyStore) describeDir(dirName string) ([]keystore.KeyDescription, e
 	return keys, nil
 }
 
+func (store *KeyStore) describeOldDir(dirName string) ([]keystore.KeyDescription, error) {
+	files, err := store.fs.ReadDir(dirName)
+	if err != nil {
+		return nil, err
+	}
+
+	keys := make([]keystore.KeyDescription, 0, len(files))
+	for _, fileInfo := range files {
+		if fileInfo.IsDir() && fileInfo.Name() == ".poison_key" {
+			//recursive read to scan poison directory
+			poisonKeys, err := store.describeOldDir(filepath.Join(dirName, fileInfo.Name()))
+			if err != nil {
+				return nil, err
+			}
+			keys = append(keys, poisonKeys...)
+
+			continue
+		}
+
+		if !fileInfo.IsDir() || !strings.HasSuffix(fileInfo.Name(), "old") {
+			continue
+		}
+
+		// remove old suffix from dir name and path it to get KeyDescription
+		fileName := strings.TrimSuffix(fileInfo.Name(), ".old")
+
+		description, err := DescribeKeyFile(fileName)
+		if err != nil {
+			return nil, err
+		}
+		if description.Purpose == keystore.PurposeLegacy {
+			log.WithField("ID", description.ID).Warn("Ignoring legacy key")
+			continue
+		}
+
+		// read files inside directory with old prefix
+		files, err = store.fs.ReadDir(filepath.Join(dirName, fileInfo.Name()))
+		if err != nil {
+			return nil, err
+		}
+
+		if len(files) == 0 {
+			log.WithField("ID", fileInfo.Name()).Warn("Contains no key files")
+			continue
+		}
+
+		for _, file := range files {
+			if file.IsDir() {
+				continue
```

Comment:
is it valid situation? maybe we should return error in such cases because it is unexpected folder structure?